File: spring-reactive-02/learn-java/src/main/java/boki/learnjava/v3_transformsequence/h_then/ThenEx1.java
package boki.learnjava.v3_transformsequence.h_then;

import boki.learnjava.utils.Logger;
import boki.learnjava.utils.TimeUtils;
import reactor.core.publisher.Mono;

import java.time.Duration;

/**
 * then 기본 개념 예제
 *  - Upstream Mono의 sequence가 종료되면, Mono<Void>를 Downstream으로 전달한다.
 */
public class ThenEx1 {
    public static void main(String[] args) {
        Mono
            .just("Hi")
            .delayElement(Duration.ofSeconds(1))
            .doOnNext(Logger::doOnNext)
            .then()
            .subscribe(
                Logger::onNext,
                Logger::onError,
                Logger::onComplete
            );

        TimeUtils.sleep(1500);
    }
}

File: spring-reactive-02/learn-kt/src/main/kotlin/boki/learnkt/v3_transformsequence/h_then/ThenEx1.kt
package boki.learnkt.v3_transformsequence.h_then

import boki.learnkt.util.Logger
import boki.learnkt.util.TimeUtils
import reactor.core.publisher.Mono
import java.time.Duration

/**
 * then 기본 개념 예제
 *  - Upstream Mono의 sequence가 종료되면, Mono<Void>를 Downstream으로 전달한다.
 */
fun main() {
    Mono
        .just("Hi")
        .delayElement(Duration.ofSeconds(1))
        .doOnNext { Logger.doOnNext(it) }
        .then()
        .subscribe(
            { Logger.onNext(it) },
            { Logger.onError(it) },
            { Logger.onComplete() }
        )

    TimeUtils.sleep(1500L)
}